From 0f14c569f1c6c8ded61da3ebfca4d94ab6b21f7a Mon Sep 17 00:00:00 2001
From: Jerome Marchand <jmarchan@redhat.com>
Date: Thu, 29 May 2014 09:38:32 -0400
Subject: [mm] avoid kABI breakage

Message-id: <1401356312-17063-6-git-send-email-jmarchan@redhat.com>
Patchwork-id: 81104
O-Subject: [RHEL7 PATCH 5/5] avoid kABI breakage
Bugzilla: 1097419
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Rafael Aquini <aquini@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla:
https://bugzilla.redhat.com/show_bug.cgi?id=1097419

The changes in swap_info_struct breaks the kABI. This is because one
argument of the (mostly unused) swap_activate() prototype of
address_space_operations is of that type. The struct address_space is
then refered to by several other structures and the modification is
eventually transmitted to a few whitelisted symbols unrelated to the
modified swap_info_struct structure.

The structure is never included directly in other structure, so hiding
the new fields at the end should be fine.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/swap.h b/include/linux/swap.h
index 225ea54..be3f325 100644
--- a/include/linux/swap.h
+++ b/include/linux/swap.h
@@ -182,9 +182,8 @@ enum {
 struct swap_info_struct {
  unsigned long flags;  /* SWP_USED etc: see above */
  signed short prio;  /* swap priority of this type */
- struct plist_node list;  /* entry in swap_active_head */
- struct plist_node avail_list; /* entry in swap_avail_head */
  signed char type;  /* strange name for an index */
+ signed char     next;           /* unused: kept for kABI */
  unsigned int max;  /* extent of the swap_map */
  unsigned char *swap_map; /* vmalloc'ed array of usage counts */
  unsigned int lowest_bit; /* index of first free in swap_map */
@@ -216,6 +215,10 @@ struct swap_info_struct {
       * swap_lock. If both locks need hold,
       * hold swap_lock first.
       */
+#ifndef __GENKSYMS__
+ struct plist_node list;  /* entry in swap_active_head */
+ struct plist_node avail_list; /* entry in swap_avail_head */
+#endif
 };
 
 /* linux/mm/workingset.c */
-- 
1.7.1